src/Storages/MergeTree/GinIndexStore.cpp

```diff
@@ -0,0 +1,447 @@
+{
+    /// Check validity of segment_id
+    auto it{ store->term_dicts.find(segment_id) };
+    if (it == store->term_dicts.cend())
+    {
+        throw Exception("Invalid segment id " + std::to_string(segment_id), ErrorCodes::LOGICAL_ERROR);
+    }
+
+    it->second->offsets.getData().clear();
+
+    /// Set file pointer of term dictionary file
+    term_dict_file_stream->seek(it->second->item_dict_start_offset, SEEK_SET);
+
+    /// Read FST size
+    size_t fst_size{0};
+    readVarUInt(fst_size, *term_dict_file_stream);
+
+    /// Read FST content
+    it->second->offsets.getData().resize(fst_size);
+    term_dict_file_stream->readStrict(reinterpret_cast<char*>(it->second->offsets.getData().data()), fst_size);
+}
+
+SegmentedPostingsListContainer GinIndexStoreReader::readSegmentedPostingsLists(const String& token)
+{
+    SegmentedPostingsListContainer container;
+    for (auto const& seg_term_dict : store->term_dicts)
+    {
+        auto segment_id = seg_term_dict.first;
+
+        auto [found, offset] = seg_term_dict.second->offsets.getOutput(token);
+        if (!found)
+            continue;
+
+        // Set postings file pointer for reading postings list
+        postings_file_stream->seek(seg_term_dict.second->postings_start_offset + offset, SEEK_SET);
+
+        // Read posting list
+        auto postings_list = GinIndexPostingsBuilder::deserialize(*postings_file_stream);
+        container[segment_id] = postings_list;
+    }
+    return container;
+}
+
+PostingsCachePtr GinIndexStoreReader::loadPostingsIntoCache(const std::vector<String>& terms)
+{
+    auto postings_cache = std::make_shared<PostingsCache>();
+    for (const auto& term : terms)
+    {
+        // Make sure don't read for duplicated terms
+        if (postings_cache->find(term) != postings_cache->cend())
+            continue;
+
+        auto container = readSegmentedPostingsLists(term);
+        (*postings_cache)[term] = container;
+    }
+    return postings_cache;
+}
+
+GinIndexStoreFactory& GinIndexStoreFactory::instance()
+{
+    static GinIndexStoreFactory instance;
+    return instance;
+}
+
+GinIndexStorePtr GinIndexStoreFactory::get(const String& name, DataPartStoragePtr storage_)
+{
+    const String& part_path = storage_->getRelativePath();
+    std::lock_guard lock(stores_mutex);
+    String key = name + String(":")+part_path;
+
+    GinIndexStores::const_iterator it = stores.find(key);
+
+    if (it == stores.cend())
+    {
+        GinIndexStorePtr store = std::make_shared<GinIndexStore>(name);
+        store->SetStorage(storage_);
+        if (!store->exists())
+            return nullptr; //gin index was not generated for the part due to no data or the length of column data is less than ngramsize
+
+        GinIndexStoreReader reader(store);
+        reader.readSegments();
+
+        for (size_t seg_index = 0; seg_index < store->getSegmentNum(); ++seg_index)
+        {
+            reader.readTermDictionary(seg_index);
```

Comment:
Since term dictionaries has to be loaded together to avoid wrong query results(suppose there is another query, which might use the same set of term dictionaries), it is kinds of  index "warm-up" which may take some time, this is expected for user. So let's leave it as IMHO.